File: 55-jump-game/55-jump-game.java
class Solution {
    Boolean[] memo;
    public boolean canJump(int[] nums) {
        int n = nums.length;
        memo = new Boolean[n];
        return jump(nums, 0);
    }
    public boolean jump(int[] nums, int start) {
        if(start == nums.length - 1) return true;
        if(memo[start] != null) return memo[start];
        int next = Math.min(nums.length - 1, start + nums[start]);
        for(int i = start + 1; i<= next; i++) {
            if(jump(nums, i))
                return memo[start] = true;
        }
        return memo[start] = false;
    }
}

File: 55-jump-game/55-jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        dp = [0] * n
        dp[0] = nums[0]
        for i in range(1,n):
            if dp[i - 1] < i: return False
            ###动态转换 不太好理解
            dp[i] = max(dp[i - 1], nums[i] + i)
            if dp[i] >= n - 1: return True
        return True if dp[n - 1] >= n - 1 else False